text/3621-derive-smart-pointer.md

```diff
@@ -337,6 +349,47 @@ you can use them for dynamic dispatch.
 As seen in the `Rc` example, the macro needs to be usable even if the pointer
 is `NonNull<RcInner<T>>` (as opposed to `NonNull<T>`).
 
+## `PinCoerceUnsized`
+
+The standard library defines the following unstable trait:
+```rust
+/// Trait that indicates that this is a pointer or a wrapper for one, where
+/// unsizing can be performed on the pointee when it is pinned.
+///
+/// # Safety
+///
+/// If this type implements `Deref`, then the concrete type returned by `deref`
+/// and `deref_mut` must not change without a modification. The following
+/// operations are not considered modifications:
+///
+/// * Moving the pointer.
+/// * Performing unsizing coercions on the pointer.
+/// * Performing dynamic dispatch with the pointer.
+/// * Calling `deref` or `deref_mut` on the pointer.
+///
+/// The concrete type of a trait object is the type that the vtable corresponds
+/// to. The concrete type of a slice is an array of the same element type and
+/// the length specified in the metadata. The concrete type of a sized type
+/// is the type itself.
+pub unsafe trait PinCoerceUnsized<U>: CoerceUnsized<U> {}
```

Comment:
you could just use a new trait `PinSmartPointer`, which, because it's named after `derive(SmartPointer)` has no naming issues and because it has no type parameters won't conflict with any future shape of `CoerceUnsized`, so I think that means `PinSmartPointer` can be stabilized without issues. If Rust later wants to split `PinSmartPointer`, that can be done with blanket impls for whatever new traits are used, or, a trait alias (if that works by then).

```rust
pub unsafe trait PinSmartPointer {}

impl<T, U> CoerceUnsized<Pin<U>> for Pin<T>
where
    T: CoerceUnsized<U> + PinSmartPointer,
    U: PinSmartPointer,
{
}

impl<T, U> DispatchFromDyn<Pin<U>> for Pin<T>
where
    T: PinSmartPointer + DispatchFromDyn<U>,
    U: PinSmartPointer,
{
}
```

Reply:
That seems reasonable enough to me.